File: cocos-creator-project/assets/scripts/game/TrackManager.ts
import { TrackData, TrackClassification, GameConstants } from './GameTypes';
import { Aircraft } from './Aircraft';
import { IFFSystem } from './IFFSystem';

// ============================================================================
// Track Manager - assigns and manages track IDs for all radar contacts
// ============================================================================

interface TrackEntry {
    trackId: number;
    aircraft: Aircraft;
    cachedData: TrackData;
}

export class TrackManager {
    private tracks: TrackEntry[] = [];
    private nextTrackId: number = 1;
    private iffSystem: IFFSystem = new IFFSystem();

    addTrack(aircraft: Aircraft): number {
        if (!aircraft) return -1;

        const id = this.nextTrackId++;
        aircraft.setTrackId(id);

        this.tracks.push({
            trackId: id,
            aircraft,
            cachedData: aircraft.getTrackData(),
        });

        this.iffSystem.interrogate(aircraft);
        return id;
    }

    removeTrack(trackId: number): void {
        this.tracks = this.tracks.filter((e) => e.trackId !== trackId);
    }

    update(dt: number): void {
        this.iffSystem.update(dt);

        this.tracks = this.tracks.filter((entry) => {
            if (!entry.aircraft || !entry.aircraft.isAlive() ||
                entry.aircraft.getRange() > GameConstants.RADAR_MAX_RANGE_KM) {
                return false;
            }
            entry.cachedData = entry.aircraft.getTrackData();
            return true;
        });
    }

    getAllTracks(): TrackData[] {
        return this.tracks.map((e) => e.cachedData);
    }

    getTrack(trackId: number): TrackData | null {
        const entry = this.tracks.find((e) => e.trackId === trackId);
        return entry ? entry.cachedData : null;
    }

    getAircraftByTrackId(trackId: number): Aircraft | null {
        const entry = this.tracks.find((e) => e.trackId === trackId);
        return entry ? entry.aircraft : null;
    }

    getHostileTrackIds(): number[] {
        return this.tracks
            .filter((e) => e.cachedData.classification === TrackClassification.HOSTILE)
            .map((e) => e.trackId);
    }

    getActiveTrackCount(): number { return this.tracks.length; }

    getHostileCount(): number {
        return this.tracks.filter(
            (e) => e.cachedData.classification === TrackClassification.HOSTILE,
        ).length;
    }

    getFriendlyCount(): number {
        return this.tracks.filter(
            (e) => e.cachedData.classification === TrackClassification.FRIENDLY,
        ).length;
    }

    getIFFSystem(): IFFSystem { return this.iffSystem; }

    reset(): void {
        this.tracks = [];
        this.nextTrackId = 1;
    }
}

File: src/TrackManager.cpp
#include "TrackManager.h"
#include <algorithm>

TrackManager::TrackManager()
    : nextTrackId_(1)
{
}

TrackManager::~TrackManager() {}

int TrackManager::addTrack(Aircraft* aircraft)
{
    if (!aircraft) return -1;

    int id = nextTrackId_++;
    aircraft->setTrackId(id);

    TrackEntry entry;
    entry.trackId = id;
    entry.aircraft = aircraft;
    entry.cachedData = aircraft->getTrackData();

    tracks_.push_back(entry);

    // Begin IFF interrogation
    iffSystem_.interrogate(aircraft);

    return id;
}

void TrackManager::removeTrack(int trackId)
{
    tracks_.erase(
        std::remove_if(tracks_.begin(), tracks_.end(),
            [trackId](const TrackEntry& e) { return e.trackId == trackId; }),
        tracks_.end());
}

void TrackManager::update(float dt)
{
    // Update IFF system
    iffSystem_.update(dt);

    // Update all track data and remove dead/out-of-range tracks
    auto it = tracks_.begin();
    while (it != tracks_.end()) {
        if (!it->aircraft || !it->aircraft->isAlive() ||
            it->aircraft->getRange() > GameConstants::RADAR_MAX_RANGE_KM) {
            it = tracks_.erase(it);
        } else {
            updateTrackData(*it);
            ++it;
        }
    }
}

void TrackManager::updateTrackData(TrackEntry& entry)
{
    if (entry.aircraft) {
        entry.cachedData = entry.aircraft->getTrackData();
    }
}

std::vector<TrackData> TrackManager::getAllTracks() const
{
    std::vector<TrackData> result;
    result.reserve(tracks_.size());
    for (const auto& entry : tracks_) {
        result.push_back(entry.cachedData);
    }
    return result;
}

TrackData* TrackManager::getTrack(int trackId)
{
    for (auto& entry : tracks_) {
        if (entry.trackId == trackId) {
            return &entry.cachedData;
        }
    }
    return nullptr;
}

Aircraft* TrackManager::getAircraftByTrackId(int trackId) const
{
    for (const auto& entry : tracks_) {
        if (entry.trackId == trackId) {
            return entry.aircraft;
        }
    }
    return nullptr;
}

std::vector<int> TrackManager::getHostileTrackIds() const
{
    std::vector<int> result;
    for (const auto& entry : tracks_) {
        if (entry.cachedData.classification == TrackClassification::HOSTILE) {
            result.push_back(entry.trackId);
        }
    }
    return result;
}

int TrackManager::getActiveTrackCount() const
{
    return static_cast<int>(tracks_.size());
}

int TrackManager::getHostileCount() const
{
    int count = 0;
    for (const auto& entry : tracks_) {
        if (entry.cachedData.classification == TrackClassification::HOSTILE) {
            ++count;
        }
    }
    return count;
}

int TrackManager::getFriendlyCount() const
{
    int count = 0;
    for (const auto& entry : tracks_) {
        if (entry.cachedData.classification == TrackClassification::FRIENDLY) {
            ++count;
        }
    }
    return count;
}

void TrackManager::reset()
{
    tracks_.clear();
    nextTrackId_ = 1;
}
